### ISSUE TITLE
operatorfunc with lua-function

### ISSUE THREAD

**Issue by AckslD** [neovim/neovim#14157]
- `nvim --version`: v0.5.0-dev+1143-gdc8273f2f
- `vim -u DEFAULTS` (version: ) behaves differently? No
- Operating system/version: 5.11.5-arch1-1
- Terminal name/version: kitty 0.19.3
- `$TERM`: xterm-kitty

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
:set operatorfunc=v:lua.require('mymod').myfunc
g@
```

### Actual behaviour
Nothing happens

### Expected behaviour
The function `myfunc` in the module `mymod` to be called.

### Possible reason
So I was trying to see why this is happening. Looking at the source code there is a [check](https://github.com/neovim/neovim/blob/master/src/nvim/eval.c#L1042) if a function starts with `v:lua.`. But then in [`check_luafunc_name`](https://github.com/neovim/neovim/blob/master/src/nvim/eval.c#L8501) it is asserted that all characters are `alphanum` or `_`/`.`. This check then fails I guess since the operatorfunc in this case also has `'`-characters. I haven't look much further yet to see what a solution might be but wanted to check if someone already knows since I'm not very familiar with the codebase yet.

**Comment by IndianBoy42** [comment]
Even more ideally, just like keymaps and autocommands are going in the direction of accepting lua functions directly. It would be great if operatorfunc could do the same. Then no global functions or even `require` based exporting would be required.

Maybe something like `vim.operatorfunc.set(<function>)`. You also generally want to stash the previous operatorfunc somewhere so that we can nest operators, `operatorfunc.set` probably should do that to. Maybe we'd call it `operatorfunc.append/stack` then

**Comment by zdcthomas** [comment]
I would love to try to contribute here, but it's my first time in the codebase. Do you have a suggestion on where I might look to get started on this?
I'd imagine this would be somewhat similar to using lua functions in keymap.set, right?

**Comment by zeertzjq** [comment]
No. I guess this more similar to Lua usercommands or autocommands.

**Comment by zdcthomas** [comment]
Ahh ok cool, any pointers on learning more about their implementation before I dive in?

**Comment by tiagovla** [comment]
Since there was no activity on this. Can I create a draft PR?
I got this working so far:
```lua
local counter = 0
local function dot_repeat(motion)
    if motion == nil then
        vim.api.nvim_set_operatorfunc(dot_repeat)
        return "g@"
    end
    print("counter:", counter, "motion:", motion)
    counter = counter + 1
end
vim.keymap.set("n", "gt", dot_repeat, { expr = true })
```

**Comment by mroavi** [comment]
Could you include a reference to the PR?

**Comment by tiagovla** [comment]
> Could you include a reference to the PR?

https://github.com/neovim/neovim/pull/20187

**Comment by zeertzjq** [comment]
In v0.9.0-dev there is a way to set `*func` options without adding global functions:

`'operatorfunc'`:
```lua
local set_opfunc = vim.fn[vim.api.nvim_exec([[
  func s:set_opfunc(val)
    let &opfunc = a:val
  endfunc
  echo get(function('s:set_opfunc'), 'name')
]], true)]
```

`'tagfunc'` for current buffer:
```lua
local set_tagfunc = vim.fn[vim.api.nvim_exec([[
  func s:set_tagfunc(val)
    let &l:tagfunc = a:val
  endfunc
  echo get(function('s:set_tagfunc'), 'name')
]], true)]
```

**Comment by zeertzjq** [comment]
Related: #18132

**Comment by yyy33** [comment]
`v:lua.require'mymod'.myfunc`
is this work?

**Comment by zeertzjq** [comment]
> `v:lua.require'mymod'.myfunc`
> is this work?

Yes, this has always worked.

**Comment by AckslD** [comment]
Just in case anyone else stumbles on this, at least from what I can tell, neither:
* `v:lua.require('mymod')`
* `v:lua.require"mymod"`
works, but
* `v:lua.require'mymod'`
does. So no brackets and single-quotes are required.

**Comment by justinmk** [comment]
That's (now) mentioned at `:help v:lua-call`:

> > Note: Only single quote form without parens is allowed. Using `require"mypack"` or `require('mypack')` as prefixes do NOT work (the latter is still valid as a function call of itself, in case require returns a useful value).

**Comment by justinmk** [comment]
This limitation is documented and I don't think there's a path to fixing it. Instead the ideas in https://github.com/neovim/neovim/issues/11306 would be an alternative. See also https://github.com/neovim/neovim/issues/18132

**Comment by mangelozzi** [comment]
> ```lua
> local set_opfunc = vim.fn[vim.api.nvim_exec([[
>   func s:set_opfunc(val)
>     let &opfunc = a:val
>   endfunc
>   echon get(function('s:set_opfunc'), 'name')
> ]], true)]
> ```

If I may ask, what is `set_opfunc` in this example, and how does it relate to `vim.go.operatorfunc`? Not sure how to use above snippet, which was the original function, and which is now assignable to `vim.go.operatorfunc`?

**Comment by zeertzjq** [comment]
If you pass a Lua function to `set_opfunc()`, it will set the operatorfunc to that function.

**Comment by mangelozzi** [comment]
> If you pass a Lua function to `set_opfunc()`, it will set the operatorfunc to that function.

Thank you so much!
